File: crabzilla-example1/crabzilla-example1-core/src/main/java/crabzilla/example1/aggregates/Customer.java
package crabzilla.example1.aggregates;

import crabzilla.example1.services.SampleInternalService;
import crabzilla.model.AggregateRoot;
import crabzilla.model.DomainEvent;
import lombok.Value;
import lombok.experimental.Wither;

import java.util.List;

import static java.util.Collections.singletonList;
import static org.apache.commons.lang3.Validate.validState;

@Value
@Wither
public class Customer implements AggregateRoot {

  transient SampleInternalService service;

  CustomerData.CustomerId id;
  String name;
  boolean isActive;
  String reason;

  List<DomainEvent> create(CustomerData.CustomerId id, String name) {

    validState(this.id == null, "customer already created");

    return singletonList(new CustomerData.CustomerCreated(id, name));
  }

  List<DomainEvent> activate(String reason) {

    return singletonList(new CustomerData.CustomerActivated(reason, service.now()));
  }

  List<DomainEvent> deactivate(String reason) {

    return singletonList(new CustomerData.CustomerDeactivated(reason, service.now()));
  }

  public static Customer of(CustomerData.CustomerId id, String name, boolean isActive, String reason) {

    return new Customer(null, id, name, isActive, reason);
  }

}

File: crabzilla-kotlin-example1/crabzilla-kotlin-example1-core/src/main/java/io/github/crabzilla/example1/customer/Customer.kt
package io.github.crabzilla.example1.customer

import io.github.crabzilla.example1.SampleInternalService
import io.github.crabzilla.example1.eventsOf
import io.github.crabzilla.model.Aggregate
import io.github.crabzilla.model.DomainEvent
import io.github.crabzilla.model.EntityId

// tag::aggregate[]

data class Customer(val _id: CustomerId? = null,
                    val name: String? = null,
                    val isActive: Boolean? = false,
                    val reason: String? = null,
                    val sampleInternalService: SampleInternalService) : Aggregate {

  override fun getId(): EntityId {
    return _id!!
  }

  internal fun create(id: CustomerId, name: String): List<DomainEvent> {
    require(this._id == null, { "customer already created" })
    return eventsOf(CustomerCreated(id, name))
  }

  internal fun activate(reason: String): List<DomainEvent> {
    require(this._id != null, { "customer must exists" })
    return eventsOf(CustomerActivated(reason, sampleInternalService.now()))
  }

  internal fun deactivate(reason: String): List<DomainEvent> {
    require(this._id != null, { "customer must exists" })
    return eventsOf(CustomerDeactivated(reason, sampleInternalService.now()))
  }

}


// end::aggregate[]
